File: webaruhaz/erettsegi/2020_maj/informatikai_ism_megoldasok_k2011/2. szudoku/java megoldasok/sudoku_cli_java/sudokucli.java
import java.io.*;
import java.util.*;

class Feladvany {
    public String kezdo;
    private int meret;
    public int getMeret(){
        return meret;
    }
    public void setMeret(int ertek){
        meret = ertek;
    }

    public Feladvany(String sor) {
        kezdo = sor;
        meret = (int) (Math.sqrt(sor.length()));
    }

    public void kirajzol() {
        for (int i = 0; i < kezdo.length(); i++) {
            if (kezdo.charAt(i) == '0') {
                System.out.print(".");
            } else {
                System.out.print(kezdo.charAt(i));
            }
            if (i % meret == meret - 1) {
                System.out.println();
            }
        }
    }
}

public class SudokuCLI {

    public static void main(String[] args) {
        //3.feladat
        List<Feladvany> feladvanyok = new ArrayList<Feladvany>();
        File inputFile = new File("./feladvanyok.txt");
        try (Scanner scanner = new Scanner(inputFile)) {
            while (scanner.hasNextLine()) {
                String aktualisSor = scanner.nextLine();
                Feladvany sorsolas = new Feladvany(aktualisSor);
                feladvanyok.add(sorsolas);
            }
        } catch (FileNotFoundException exception) {
            System.err.print("Fájl nem található!");
        }
        System.out.println("2.feladat: Beolvasva " + feladvanyok.size() + " feladvány.");

        //4.feladat
        int beolvasottMeret;
        Scanner consoleScanner = new Scanner(System.in);
        do {
            System.out.println("3.feladat: Kérem a feladvány méretét [4..9]:");
            beolvasottMeret = Integer.parseInt(consoleScanner.next());
        } while (beolvasottMeret < 4 || beolvasottMeret > 9);
        int meretDb = 0;
        for (Feladvany feladvany : feladvanyok) {
            if (feladvany.getMeret() == beolvasottMeret) {
                meretDb++;
            }
        }
        System.out.println(beolvasottMeret + "x" + beolvasottMeret + " méretű feladványból " + meretDb + "darab van tárolva.");

        //5.feladat
        List<Feladvany> szurtLista = new ArrayList<Feladvany>();
        Random rand = new Random();

        for (Feladvany feladvany : feladvanyok) {
            if (feladvany.getMeret() == beolvasottMeret) {
                szurtLista.add(feladvany);
            }
        }

        int veletlenSzam = rand.nextInt(szurtLista.size());
        Feladvany kivalasztott = szurtLista.get(veletlenSzam);
        System.out.println("4.feladat: A kiválasztott feladvány: " + kivalasztott.kezdo);

        //6.feladat
        double uresDb = 0;
        for (char c : kivalasztott.kezdo.toCharArray()) {
            if (c != '0') {
                uresDb++;
            }
        }
        double megoldas = (uresDb / (kivalasztott.getMeret() * kivalasztott.getMeret())) * 100;
        int megoldasKerekitve =  (int) Math.round(megoldas);

        System.out.println("5.feladat: A feladvány kitöltöttsége: " + megoldasKerekitve + "%");

        //7.feladat
        System.out.println("6.feladat: A feladvány kirajzolva:");
        kivalasztott.kirajzol();

        //8.feladat
        String fajlnev = "sudoku" + beolvasottMeret + ".txt";
        try {
            BufferedWriter writer = new BufferedWriter(new FileWriter(fajlnev, false));
            for (Feladvany feladvany : feladvanyok) {
                if (feladvany.getMeret() == beolvasottMeret) {
                    writer.write(feladvany.kezdo);
                    writer.newLine();
                }
            }
            writer.close();
        } catch (IOException e) {
            System.err.print("Hiba a fájlírás során!");
        }
    }
}

File: webaruhaz/erettsegi/2020_maj/informatikai_ism_megoldasok_k2011/2. szudoku/c# megoldasok/sudoku_cli_csharp/sudokucli.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

namespace sudokuCLI
{
    class Feladvany
    {
        public string Kezdo { get; private set; }
        public int Meret { get; private set; }

        public Feladvany(string sor)
        {
            Kezdo = sor;
            Meret = Convert.ToInt32(Math.Sqrt(sor.Length));
        }

        public void Kirajzol()
        {
            for (int i = 0; i < Kezdo.Length; i++)
            {
                if (Kezdo[i] == '0')
                {
                    Console.Write(".");
                }
                else
                {
                    Console.Write(Kezdo[i]);
                }
                if (i % Meret == Meret - 1)
                {
                    Console.WriteLine();
                }
            }
        }
    }
    class sudokuCLI
    {
        static void Main(string[] args)
        {
            //3.feladat
            List<Feladvany> feladvanyok = new List<Feladvany>();
            foreach (string Seged in File.ReadAllLines("feladvanyok.txt"))
            {
                feladvanyok.Add(new Feladvany(Seged));
            }
            Console.WriteLine("3. feladat: Beolvasva {0} feladvány", feladvanyok.Count);

            //4.feladat
            int meret;
            do
            {
                Console.Write("\n4. feladat: Kérem a feladvány méretét [4..9]: ");
                meret = Convert.ToInt32(Console.ReadLine());
            } while (meret < 4 || meret > 9);
            int meretDb = 0;
            foreach (var Seged in feladvanyok)
            {
                if (Seged.Meret == meret)
                {
                    meretDb++;
                }
            }
            Console.WriteLine("{0}x{0} méretű feladványból {1} darab van tárolva", meret, meretDb);

            //5.feladat
            Random Veletlen = new Random();
            int kivalasztott;
            do
            {
                kivalasztott = Veletlen.Next(0, feladvanyok.Count);
            } while (feladvanyok[kivalasztott].Meret != meret);
            Console.WriteLine("\n5. feladat: A kiválasztott feladvány:\n{0}", feladvanyok[kivalasztott].Kezdo);

            //6.feladat
            int toltottDb = 0;
            foreach (char szamJegy in feladvanyok[kivalasztott].Kezdo)
            {
                if (szamJegy != '0')
                {
                    toltottDb++;
                }
            }
            double szazalekErtek = Math.Round((double)100 * toltottDb / feladvanyok[kivalasztott].Kezdo.Length, 0, MidpointRounding.AwayFromZero);
            Console.WriteLine("\n6. feladat: A feladvány kitöltöttsége: {0}%", szazalekErtek);

            //7.feladat
            Console.WriteLine("\n7. feladat: A feladvány kirajzolva:");
            feladvanyok[kivalasztott].Kirajzol();

            //8.feladat
            StreamWriter Ki = new StreamWriter("sudoku" + meret + ".txt");
            foreach (var seged in feladvanyok)
            {
                if (seged.Meret == meret)
                {
                    Ki.WriteLine(seged.Kezdo);
                }
            }
            Ki.Close();
            Console.WriteLine("\n8. feladat: sudoku{0}.txt állomány {1} darab feladvánnyal létrehozva", meret, meretDb);

            Console.ReadLine();
        }
    }
}
